### PR TITLE
refactor Ackermann mix function for DRYness and readability

### PR DESCRIPTION
**Pull request by robberwick**
refactor some of the logic in `Ackermann::mix` so that there is less repetition, and so that it is more readable. Also do some drive-by code improvements while I'm there (thanks clang-tidy)

### CODE REVIEW

File: libs/mixer/src/ackermann_mixer.cpp [hunk 1]
```diff
@@ -21,7 +21,12 @@ int sign(T value) {
 }
 
 namespace MIXER {
-AckermannMixer::AckermannMixer(float track, float base, float angle) : wheelTrack(track), wheelBase(base), maxSteeringAngle(angle) {} 
+
+    AckermannMixer::AckermannMixer(float track, float base, float angle) : wheelTrack(track), wheelBase(base),
+                                                                           maxSteeringAngle(angle) {
+        // initialise the turn radius to zero
+        turnRadius = 0.0;
```
Comment (markmellors) [inline libs/mixer/src/ackermann_mixer.cpp:29]: I thought this was going to be initialised to null? It's a minor thing, but if anything queries turnRadius before we've sent a move command, it'll think we're turning on the spot.  One potential consequence of that is the initial/resting state might be with the steering wheels toed in, whereas I think we'd prefer to default to straight ahead (assuming that the first move is more likely to be forwards, than a turn on the spot)

File: libs/mixer/src/ackermann_mixer.cpp [hunk 2]
```diff
@@ -28,77 +29,143 @@ namespace MIXER {
...
+            // if the turn radius is infinite then we're not turning, so all wheels travel
+            // at the desired forwards speed and the steerable wheels point forwards
+            result.frontLeftSpeed = velocity;
+            result.frontRightSpeed = -velocity;
+            result.rearLeftSpeed = velocity;
+            result.rearRightSpeed = -velocity;
+            result.frontLeftAngle = 0;
+            result.frontRightAngle = 0;
+        } else {
+            const float halfWheelTrack = wheelTrack / 2;
```
Comment (markmellors) [inline libs/mixer/src/ackermann_mixer.cpp:57]: we could put this parameter in drivetrain_config, as it won't change during a run

File: libs/mixer/src/ackermann_mixer.cpp [hunk 3]
```diff
@@ -28,77 +29,143 @@ namespace MIXER {
...
+
+            // calculate the x component of the turn radius of each wheel
+            // where x is left/right and y is direction of travel
+            const float leftWheelTurnRadius = turnRadius - halfWheelTrack;
+            const float rightWheelTurnRadius = turnRadius + halfWheelTrack;
+
+
+            if (velocity == 0) {
+                // if we're only turning, the speeds are symmetrical and just depends on the turn rate
+                result.frontRightSpeed = result.frontLeftSpeed = -angularVelocity * std::sqrt((halfWheelTrack) * (halfWheelTrack) + wheelBase * wheelBase);
```
Comment (markmellors) [inline libs/mixer/src/ackermann_mixer.cpp:67]: we could also put the sqrt term in drivetrain_config, as it won't change during a run. its the distance (hypotenuse) from the centre of the rear axle to the front wheels

File: libs/mixer/src/ackermann_mixer.cpp [hunk 4]
```diff
@@ -28,77 +29,143 @@ namespace MIXER {
         turnRadius = 0.0;
     }
 
-AckermannOutput AckermannMixer::mix(float velocity, float angularVelocity){
-    // function takes desired forward speed ("throttle") in mm/s and turn rate in radians/sec
-    // and outputs individual wheel speeds in mm/sec and turn angle of steerable wheels in radians
+    AckermannOutput AckermannMixer::mix(float velocity, float angularVelocity) {
+        // function takes desired forward speed ("throttle") in mm/s and turn rate in radians/sec
+        // and outputs individual wheel speeds in mm/sec and turn angle of steerable wheels in radians
```
Comment (markmellors) [inline libs/mixer/src/ackermann_mixer.cpp:34]: old, old commemt is incorrect, output is in m/s

**Comment by markmellors** [comment]
code tested on test bed - as far as I can tell, it works the same as before 👍
left some comments on minor points

File: libs/mixer/src/ackermann_mixer.cpp (hunk 2 above)
Comment (robberwick) [reply libs/mixer/src/ackermann_mixer.cpp:57]: an excellent point :+1:

File: libs/mixer/src/ackermann_mixer.cpp (hunk 3 above)
Comment (robberwick) [reply libs/mixer/src/ackermann_mixer.cpp:67]: also true :+1:

File: libs/mixer/src/ackermann_mixer.cpp (hunk 4 above)
Comment (robberwick) [reply libs/mixer/src/ackermann_mixer.cpp:34]: ok, I'll update it to `and outputs individual wheel speeds in m/s and turn angle of steerable wheels in radians`

File: libs/mixer/src/ackermann_mixer.cpp (hunk 1 above)
Comment (robberwick) [reply libs/mixer/src/ackermann_mixer.cpp:29]: does this actually matter in practice? when `mix()` is called, we set `turnRadius` based on `angularVelocity`, with no regard to its current value. We never _don't_ set it, so it'll never use the value that's set in the constructor, AFAICT. The setting here is only to make `clang-tidy` happier, and would work exactly the same if it was uninitialised - as is borne out by the fact that it worked exactly the same before this change was made.

**Pull request merged**